executor/infoschema_reader.go

```diff
@@ -2167,6 +2135,109 @@ func (e *stmtSummaryTableRetriever) retrieve(ctx context.Context, sctx sessionct
 	return rows, nil
 }
 
+type tidbTrxTableRetriever struct {
+	dummyCloser
+	batchRetrieverHelper
+	table       *model.TableInfo
+	columns     []*model.ColumnInfo
+	txnInfo     []*txninfo.TxnInfo
+	initialized bool
+}
+
+func (e *tidbTrxTableRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+
+	if !e.initialized {
+		e.initialized = true
+
+		sm := sctx.GetSessionManager()
+		if sm == nil {
+			e.retrieved = true
+			return nil, nil
+		}
+
+		loginUser := sctx.GetSessionVars().User
+		hasProcessPriv := hasPriv(sctx, mysql.ProcessPriv)
+		infoList := sm.ShowTxnList()
+		e.txnInfo = make([]*txninfo.TxnInfo, 0, len(infoList))
+		for _, info := range infoList {
+			// If you have the PROCESS privilege, you can see all running transactions.
+			// Otherwise, you can see only your own transactions.
+			if !hasProcessPriv && loginUser != nil && info.Username != loginUser.Username {
+				continue
+			}
+			e.txnInfo = append(e.txnInfo, info)
+		}
+
+		e.batchRetrieverHelper.totalRows = len(e.txnInfo)
+		e.batchRetrieverHelper.batchSize = 1024
+	}
+
+	var err error
+	var instanceAddr string
+	switch e.table.Name.O {
+	case infoschema.ClusterTableTiDBTrx:
+		instanceAddr, err = infoschema.GetInstanceAddr(sctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var res [][]types.Datum
+	err = e.nextBatch(func(start, end int) error {
+		// Before getting rows, collect the SQL digests that needs to be retrieved first.
+		var sqlRetriever *SQLDigestTextRetriever
+		for _, c := range e.columns {
+			if c.Name.O == txninfo.CurrentSQLDigestTextStr {
+				if sqlRetriever == nil {
+					sqlRetriever = NewSQLDigestTextRetriever()
+				}
+
+				for i := start; i < end; i++ {
+					sqlRetriever.SQLDigestsMap[e.txnInfo[i].CurrentSQLDigest] = ""
+				}
+			}
+		}
+		// Retrieve the SQL texts if necessary.
+		if sqlRetriever != nil {
+			err1 := sqlRetriever.RetrieveLocal(ctx, sctx)
+			if err1 != nil {
+				return errors.Trace(err1)
+			}
+		}
+
+		res = make([][]types.Datum, 0, end-start)
+
+		for i := start; i < end; i++ {
+			row := make([]types.Datum, 0, len(e.columns))
+			for _, c := range e.columns {
+				if c.Name.O == util.ClusterTableInstanceColumnName {
+					row = append(row, types.NewDatum(instanceAddr))
```

Comment:
Is it expected to leave an empty string if the table is not `CLUSTER_TIDB_TRX`?

Reply:
This function handles `CLUSTER_TIDB_TRX` and `TIDB_TRX`, and the `INSTANCE` column only exists in cluster tables.